### DISCUSSION TITLE
Implementing NAME>INTERPRET

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#127] [Q&A] [answered]
How is NAME>INTERPRET typically implemented?   There does not appear to be standard word to mark a word as compile-only (similar to IMMEDIATE).   For primitives / built-ins its easy enough, but how is it done for defined words?  Do defined words have to be state smart to throw -14 or other if interpreted?

**Comment by ruv** [comment, marked as answer]
The word [`name>interpret`](https://forth-standard.org/standard/tools/NAMEtoINTERPRET) is insufficiently specified in Forth-2012.  An irony is that this word (along with `name>compile` and `find-name`) was supposed to replace `find`, since `find` is insufficiently specified, but it got into the same problem.

It is supposed that `name>interpret` returns the _xt_ (execution token) of a word, the same _xt_ that [Tick](https://forth-standard.org/standard/core/Tick) returns for the word (or that `find` returns in interpretation state).
- Hence, in the general case, to perform the interpretation semantics for a word, this _xt_ should be executed in interpretation state — albeit the glossary entry does not specify this at the moment (therefor you can get an impression that you can execute this _xt_ in compilation state as well as in interpretation state and get the same result — but this is incorrect, since the returned _xt_ may identify state-dependent execution semantics).

In classic single-xt systems this word corresponds to the word `name> ( nt -- xt )`  that just fetches the _xt_ from a word header.

In dual-xt systems this word returns the primary _xt_ for an _nt_, which identifies the execution semantics for the word identified by the _nt_.
- Some people tend to claim that this _xt_ also identifies the interpretation semantics for the word (and the name reflects this idea), but I believe this is incorrect in the general case.


>  Do defined words have to be state smart to throw -14 or other if interpreted?

In a standard program, user-defined words should somehow analyze state to throw an exception only in interpretation state. Classic single-xt systems do the same. Usually they have a helper word for that, something like:
```
: ?comp ( -- ) \ throw code -14 if the system is not in compilation state
  state @ 0= -14 and throw
;
```

**Comment by ruv** [comment]
I have prepared a [proposal](https://forth-standard.org/proposals/name-interpret-wording?hideDiff#reply-1101) to better specify what the results of `name>interpret` mean.

NB: the standard defines default interpretation semantics (see [3.4.3.2](https://forth-standard.org/standard/usage#usage:interpret)), so they are undefined by the standard only if this fact is explicitly declared as "_Interpretation: Interpretation semantics for this word are undefined_".
